dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordMapper.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.SemanticKernel.Data;
+using Pinecone;
+
+namespace Microsoft.SemanticKernel.Connectors.Pinecone;
+
+/// <summary>
+/// Mapper between a Pinecone record and the consumer data model that uses json as an intermediary to allow supporting a wide range of models.
+/// </summary>
+/// <typeparam name="TRecord">The consumer data model to map to or from.</typeparam>
+internal sealed class PineconeVectorStoreRecordMapper<TRecord> : IVectorStoreRecordMapper<TRecord, Vector>
+    where TRecord : class
+{
+    /// <summary>A set of types that a key on the provided model may have.</summary>
+    private static readonly HashSet<Type> s_supportedKeyTypes = [typeof(string)];
+
+    /// <summary>A set of types that data properties on the provided model may have.</summary>
+    private static readonly HashSet<Type> s_supportedDataTypes =
+    [
+        typeof(bool),
+        typeof(bool?),
+        typeof(string),
+        typeof(int),
+        typeof(int?),
+        typeof(long),
+        typeof(long?),
+        typeof(float),
+        typeof(float?),
+        typeof(double),
+        typeof(double?),
+        typeof(decimal),
+        typeof(decimal?),
+        typeof(string[]),
+        typeof(List<string>),
+    ];
+
+    /// <summary>A set of types that vectors on the provided model may have.</summary>
+    private static readonly HashSet<Type> s_supportedVectorTypes = [typeof(ReadOnlyMemory<float>)];
```

Comment:
Also consider `ReadOnlyMemory<float>?` since the user may not always retrieve vectors on read, and they may therefore want to mark the vector as optional.